src/metabase/serialization/dump.clj

```diff
@@ -0,0 +1,294 @@
+          (:schema table)
+          (:name table)))
+
+(defmethod fully-qualified-name (type Field)
+  [field]
+  (str (->> field :table_id Table fully-qualified-name) "/fields/" (:name field)))
+
+(defmethod fully-qualified-name (type Metric)
+  [metric]
+  (str (->> metric :table_id Table fully-qualified-name) "/metrics/" (:name metric)))
+
+(defmethod fully-qualified-name (type Segment)
+  [segment]
+  (str (->> segment :table_id Table fully-qualified-name) "/segments/" (:name segment)))
+
+(defmethod fully-qualified-name (type Collection)
+  [collection]
+  (let [parents (->> (str/split (:location collection) #"/")
+                     rest
+                     (map #(-> % Integer/parseInt Collection :name (str "/collections")))
+                     (str/join "/")
+                     (format "%s/"))]
+    (str "/collections/collections/" parents (:name collection))))
+
+(defmethod fully-qualified-name (type Dashboard)
+  [dashboard]
+  (format "%s/dashboards/%s"
+          (or (some->> dashboard :collection_id Collection fully-qualified-name)
+              "/collections")
+          (:name dashboard)))
+
+(defmethod fully-qualified-name (type Pulse)
+  [pulse]
+  (format "%s/pulses/%s"
+          (or (some->> pulse :collection_id Collection fully-qualified-name)
+              "/collections")
+          (:name pulse)))
+
+(defmethod fully-qualified-name (type Card)
+  [card]
+  (format "%s/cards/%s"
+          (or (some->> card
+                       :dataset_query
+                       qp.util/query->source-card-id
+                       Card
+                       fully-qualified-name)
+              (some->> card
+                       :collection_id
+                       Collection
+                       fully-qualified-name)
+              "/collections")
+          (:name card)))
+
+(defmethod fully-qualified-name nil
+  [_]
+  nil)
+
+(def ^:private SegmentOrMetric
+  [(s/one (s/constrained su/KeywordOrString
+                         (comp #{:metric :segment} qp.util/normalize-token))
+          "head")
+   (s/cond-pre s/Int su/KeywordOrString)])
+
+(def ^{:arglists '([form])} entity-reference?
+  "Is given form an MBQL entity reference?"
+  (some-fn field-reference? (complement (s/checker SegmentOrMetric))))
+
+(defn- entity-reference->fully-qualified-name
+  [[op & args :as entity-reference]]
+  (case (qp.util/normalize-token op)
+    :metric        [op (fully-qualified-name (Metric (first args)))]
+    :segment       [op (fully-qualified-name (Segment (first args)))]
+    :field-id      [op (if (string? (first args))
+                         (first args)
+                         (fully-qualified-name (Field (first args))))]
+    :fk->          (into [op]
+                         (for [arg args]
+                           (if (number? arg)
+                             (Field arg)
+                             (entity-reference->fully-qualified-name arg))))
+    :field-literal entity-reference))
```

Comment:
When working with MBQL you probably want to make sure it's normalized using `mbql.normalize/normalize` (not sure if you even need to do this because everything gets normalized when it comes out of the DB now) 

Then you can know for a fact `fk->` won't have number arguments (everything will conform to the MBQL schema).

The new `metabase.mbql.util/match` family of macros could probably make this code a lot simpler

I think this code could benefit from using the new `metabase.mbql.util/match` family of macros too. My thinking behind it was since we do pattern-based matching and transformation of MBQL all over the query processor middleware it would be helpful to have a really powerful high-level set of macros for those operations that would handle all the business logic for you. The idea is to let you write concise declarative code that describes what you want to do instead of hairy code that specifies what should be done. Worth checking out, I think that could make code a lot simpler

Reply:
To build on what I was getting at, if you used `mbql.u/replace` you could take out the `:fk->` and `:field-literal` parts entirely. 

Basically what you're doing is rewriting `:metric`, `:segment`, and `:field-id` forms, right? So with `mbql.u/replace`, you would write patterns to match those three forms and rules to replace them when the patterns are matched. You wouldn't have to work about any looping. You code could basically look something like this:

```clj
(mbql.u/replace entity-reference
  [:field-id (id :guard integer?)] [:field-id (fully-qualified-name (Field id))]
  [:metric id]                     [:metric (fully-qualified-name (Metric id))]
  [:segment id]                    [:segment (fully-qualified-name (Segment id))])
```

Wayyy simpler.